### PR TITLE
KAFKA-16285: Make group metadata available when a new assignment is set

### PR DESCRIPTION
**Pull request by cadonna**
Currently, in the async Kafka consumer updates to the group metadata that are received by the heartbeat are propagated to the application thread in form of an event. Group metadata is updated when a new assignment is received. The new assignment is directly set in the subscription without sending an update event from the background thread to the application thread. That means that there might be a delay between the application thread being aware of the update to the assignment and the application thread being aware of the update to the group metadata. This delay can cause stale group metadata returned by the application thread that then causes issues when data of the new assignment is committed. A concrete example is

producer.sendOffsetsToTransaction(offsetsToCommit, groupMetadata) 

The offsets to commit might already stem from the new assignment but the group metadata might relate to the previous assignment.

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 1]
```diff
@@ -293,6 +276,23 @@ private void process(final ConsumerRebalanceListenerCallbackNeededEvent event) {
     private final AtomicLong currentThread = new AtomicLong(NO_CURRENT_THREAD);
     private final AtomicInteger refCount = new AtomicInteger(0);
 
+    private final InvalidGroupIdException invalidGroupIdException =
+        new InvalidGroupIdException("To use the group management or offset commit APIs, you must " +
+            "provide a valid " + ConsumerConfig.GROUP_ID_CONFIG + " in the consumer configuration.");
+
+    private void updateGroupMetadata(final Optional<Integer> memberEpoch, final Optional<String> memberId) {
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Please move this method. We shouldn't put private methods before the constructor.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 2]
```diff
@@ -293,6 +276,23 @@ private void process(final ConsumerRebalanceListenerCallbackNeededEvent event) {
     private final AtomicLong currentThread = new AtomicLong(NO_CURRENT_THREAD);
     private final AtomicInteger refCount = new AtomicInteger(0);
 
+    private final InvalidGroupIdException invalidGroupIdException =
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: We will have one such instance for each consumer in memory at all times. Not a biggie, but I'd avoid it.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 3]
```diff
@@ -1186,8 +1181,7 @@ public OptionalLong currentLag(TopicPartition topicPartition) {
     public ConsumerGroupMetadata groupMetadata() {
         acquireAndEnsureOpen();
         try {
-            maybeThrowInvalidGroupIdException();
-            return groupMetadata.get();
+            return groupMetadata.get().orElseThrow(() -> invalidGroupIdException);
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: I'd prefer calling `maybeThrowInvalidGroupIdException` here.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 4]
```diff
@@ -293,6 +276,23 @@ private void process(final ConsumerRebalanceListenerCallbackNeededEvent event) {
     private final AtomicLong currentThread = new AtomicLong(NO_CURRENT_THREAD);
     private final AtomicInteger refCount = new AtomicInteger(0);
 
+    private final InvalidGroupIdException invalidGroupIdException =
+        new InvalidGroupIdException("To use the group management or offset commit APIs, you must " +
+            "provide a valid " + ConsumerConfig.GROUP_ID_CONFIG + " in the consumer configuration.");
+
+    private void updateGroupMetadata(final Optional<Integer> memberEpoch, final Optional<String> memberId) {
+        groupMetadata.updateAndGet(
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Can we not just use a `volatile` variable here? We only have the background thread updating the epoch and the ID, with the exception of the shutdown procedure in the application thread. A race between background thread and application thread however doesn't really seem to be solved by using `updateAndGet` (that would require checking if the consumer group metadata is "closed" here).

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java [hunk 5]
```diff
@@ -1372,4 +1373,8 @@ public PollResult poll(final long currentTimeMs) {
         }
         return PollResult.EMPTY;
     }
+
+    List<MemberStateListener> stateListeners() {
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:1497]: // for testing only

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/RequestManagersTest.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import static org.apache.kafka.test.TestUtils.requiredConsumerConfig;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.mock;
+
+public class RequestManagersTest {
+
+    @Test
+    public void testMemberStateListenerRegistered() {
+
+        final MemberStateListener listener = (memberEpoch, memberId) -> {
```
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/RequestManagersTest.java]: nit: I would not put a new-line after this

**Review by lucasbru — COMMENTED** [review]
Thanks for the PR! I left some comments

Architecturally, this is going a bit against https://issues.apache.org/jira/browse/KAFKA-16290 which proposes propagating the subscription state via events to the application thread, which would resolve the ordering in another way. Do you have thoughts on that?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 4 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: The application thread can also update the group metadata when unsubscribing.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java (hunk 5 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:1497]: I can add it, but I really do not like those kind of inline comments in the the code.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/RequestManagersTest.java (hunk 6 above)
Comment (cadonna) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/RequestManagersTest.java]: Done

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 3 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Done

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 1 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Done

**Comment by cadonna** [comment]
> Thanks for the PR! I left some comments
> 
> Architecturally, this is going a bit against https://issues.apache.org/jira/browse/KAFKA-16290 which proposes propagating the subscription state via events to the application thread, which would resolve the ordering in another way. Do you have thoughts on that?

Did we take any decision about this architectural change proposed in https://issues.apache.org/jira/browse/KAFKA-16290?

In general I am in favor of the change, but I might miss the reason we opted to use a shared object.
 
@kirktrue @philipnee @lianetm @dajac Does anybody of you know why we chose a shared object instead of sending events between the threads?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 2 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: I removed it.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java (hunk 5 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:1497]: I do not understand why it is so important to state that this method is only used in tests on the implementation class.

Comment (lucasbru) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:1497]: I'm not saying it's "so important", but it's nice for the reader, no? I won't insist.

Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:1497]: I do not see big value for the reader. As always, I see the risk that eventually the method is not only used for testing and then we have a lying comment in the code.

**Comment by kirktrue** [comment]
> Did we take any decision about this architectural change proposed in [issues.apache.org/jira/browse/KAFKA-16290](https://issues.apache.org/jira/browse/KAFKA-16290)?

Not that I'm aware of, no. I grabbed the ticket so that it would show up on my list of things to worry about 😟

> @kirktrue @philipnee @lianetm @dajac Does anybody of you know why we chose a shared object instead of sending events between the threads?

IIRC, there were two reasons we felt comfortable using the `SubscriptionState` object directly:

1. It is heavily synchronized
2. We didn't want to add overhead for simple operations like looking up the subscribed topics, etc.

**Comment by kirktrue** [comment]
@cadonna—thanks for the PR!

I'm concerned that `MemberStateListener` mechanism has opened a path for us to sidestep the thread separation we've intentionally introduced. With this listener mechanism, the `MembershipManagerImpl` can invoke "arbitrary" code which might interact with the application thread in unexpected ways.

With all of that said, I do recognize the issue we're facing with the existing 'group metadata update event' mechanism, so it's clear we do need to tweak the design a bit. I'm just not confident we've nailed it, yet.

I will spend some time looking at this, and related, PRs and Jiras tomorrow.

**Review by lucasbru — APPROVED** [review]
LGTM, thanks!

**Comment by cadonna** [comment]
@kirktrue are you fine with merging this PR and coming back to this after 3.8?

**Comment by kirktrue** [comment]
> @kirktrue are you fine with merging this PR and coming back to this after 3.8?

Yes. I think this is an area that we need a more holistic design review, unfortunately.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java (hunk 5 above)
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:1497]: I agree.

**Review by AndrewJSchofield — APPROVED** [review]
lgtm. This design is much better. There's no need to run anything in the application thread just to update the group metadata.

**Pull request merged**